Update SNAP packaging

hurricanehrndz opened the pull request:
Enable deployment of SNAP packages from travis-ci. Benefits are as follows:

- nightly packages will be available on an Organizational controlled auto-update medium (increase feedback loop)
- stable builds the same
- other builds can be downloaded and installed from [transfer.sh](https://transfer.sh/) - [example](https://transfer.sh/UUxWV/neovim_v0.5.0-nightly-12-gb6eb063_amd64.snap)
- should increase accessibility of latest stable releases

### travis is configured as follows:
- nightly tags get push to edge channel in snap store
- stable tags get pushed to candidate channel in snap store
- commits, get pushed to transfer.sh

### requirements of team
1. Create Ubunto SSO account.
~~Neovim already has an official PPA. The same creds should be and could probably be used to create the snap store account. You can verify that [here](https://forum.snapcraft.io/t/creating-snap-store-brand-accounts/6271).~~
2. Commit secrets to Repo:
Once the snap store account has been created, one just needs to commit the secrets to this repo. You can do this by simply running [`snapcraft enable-ci travis --refresh`](https://tutorials.ubuntu.com/tutorial/continuous-snap-delivery-from-travis-ci#4). All other configuration requirements are in this commit.

If accepted I will update wiki and readme.

justinmk commented on .travis.yml:
```diff
@@ -28,8 +28,7 @@ env:
     - UBSAN_OPTIONS="print_stacktrace=1 log_path=$LOG_DIR/ubsan"
-    # Environment variables for Valgrind.
-    - VALGRIND_LOG="$LOG_DIR/valgrind-%p.log"
+    # Environment variables for Valgrind. - VALGRIND_LOG="$LOG_DIR/valgrind-%p.log"
```
?

justinmk commented on ci/snap/after_success.sh:
```diff
@@ -0,0 +1,19 @@
+
+sudo snap install "$RESULT_SNAP" --dangerous --classic
+
+/snap/bin/neovim --version
```
why is the binary named `neovim`? that is confusing, it should be named `nvim`.

hurricanehrndz replied:
@justinmk,

The snap package is called neovim, so the binary is called neovim. You can ask for an alias, but that is controlled within the snapstore. For now

```
snap alias neovim.neovim nvim
```

If you wish, I can change the package name to nvim.

justinmk commented on snap/snapcraft.yaml:
```diff
@@ -1,5 +1,8 @@
+---
+
```
what's this for?

justinmk commented on .travis.yml:
```diff
@@ -145,6 +144,33 @@ jobs:
         - CLANG_SANITIZER=TSAN
         - *common-job-env
+    - if: type != pull
+      name: snap build
```
"build" is redundant in a list of travis jobs.

```suggestion
      name: snap
```

justinmk commented:
> * other builds can be downloaded and installed from [transfer.sh](https://transfer.sh/) - [example](https://transfer.sh/UUxWV/neovim_v0.5.0-nightly-12-gb6eb063_amd64.snap)

Could mention that on [the wiki](https://github.com/neovim/neovim/wiki/Installing-Neovim#flatpak), and also add a new "snap" section there.

> * Neovim already has an official PPA. The same creds should be and could probably be used to create the snap store account. You can verify that [here](https://forum.snapcraft.io/t/creating-snap-store-brand-accounts/6271).

@jamessan @fwalch would you be able to do that?

jamessan commented:
There's no shared credentials for the PPA management.  The unstable repo just runs builds once a day and pulls from GitHub.

justinmk commented:
Not sure what that means. I will just use my own snap creds I guess.

hurricanehrndz commented:
It is okay, I will create a new account specifically named neovim and give
back the creds to you. Unless you want to create a new account?

https://snapcraft.io/docs/creating-your-developer-account

On Wed, Nov 20, 2019 at 1:12 PM Justin M. Keyes <notifications@github.com>
wrote:

> Not sure what that means. I will just use my own snap creds I guess.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/neovim/neovim/pull/11428?email_source=notifications&email_token=ABMJBTI6ODCWLJORJQOQ2ELQUWK3ZA5CNFSM4JPSAQSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEUPY7Y#issuecomment-556334207>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMJBTO6SZWGGLSBGVT7PJTQUWK3ZANCNFSM4JPSAQSA>
> .
>


-- 
Regards,

Carlos Hernandez

Commit:
snap: add "snap" job to Travis CI

hurricanehrndz commented:
I tried register the username neovim and it is already taken.

Commit:
snap: fix/update snap builds

* rename to nvim
* bump base to core18
* fix versioning, match binary
* set prefix to /usr

Commit:
snap: add desktop file, icon

Commit:
snap: add snapcraft secrets to CI

hurricanehrndz commented:
@justinmk 

This pull request is ready. I have added the required secrets on my end, the decryption key needs to be added to [travis](https://docs.travis-ci.com/user/encrypting-files/#using-openssl). I have provided back all credentials to you as well on Gitter.

hurricanehrndz commented:
PS the decryption key should be added before the pull request is accepted.

justinmk commented:
i see `-DCMAKE_BUILD_TYPE=Debug` in the snap job on travis

justinmk commented on ci/snap/deploy.sh:
```diff
@@ -0,0 +1,25 @@
+  -iter 1000 \
+  -a -d \
+  -in .snapcraft/travis_snapcraft.cfg \
+  -out .snapcraft/snapcraft.cfg -k $SNAP_SECRECT_KEY
```
why doesn't `snapcraft` do this for us? Weird that we have to maintain a bunch of scripts and `openssl` invocations, to support snap.

hurricanehrndz replied:
This is pretty standard for travis and deployments that depend on credentials. It is not isolated to snapcraft. This is one of the many ways to keeps secrets in travis. I felt this would put the least burden on the maintainers, as they only need to commit the secret key to travis.

justinmk commented:
https://tutorials.ubuntu.com/tutorial/continuous-snap-delivery-from-travis-ci#4

    travis login
    snapcraft enable-ci travis --refresh

> The dedicated credentials will be functional for exactly 1 year or until a password change on the related Ubuntu One SSO account. In order to refresh the project credentials, please run the following command:   `snapcraft enable-ci travis --refresh`

So this will break in 1 year. Not sure how to deal with all of these services that require yearly yak-shaving. Can't "schedule" GitHub issues...

> DEPRECATED: The 'snap' keyword has been replaced by 'prime'. See http://snapcraft.io/docs/deprecation-notices/dn1 for more information.

Anyways, this `snapcraft enable-ci` thing made love to the `.travis.yml` (changed a lot of unreleated stuff). Now figuring out how to reconcile that.

justinmk commented on ci/snap/deploy.sh:
```diff
@@ -0,0 +1,25 @@
+  -pbkdf2 \
+  -iter 1000 \
+  -a -d \
+  -in .snapcraft/travis_snapcraft.cfg \
```
can we move this to `snap/`? We already have 2 folders for snap (`snap/` and `ci/snap/`), adding a third is pretty annoying.

justinmk commented on ci/snap/deploy.sh:
```diff
@@ -0,0 +1,25 @@
+  -iter 1000 \
+  -a -d \
+  -in .snapcraft/travis_snapcraft.cfg \
+  -out .snapcraft/snapcraft.cfg -k $SNAP_SECRECT_KEY
```
`SNAP_SECRECT_KEY` is misspelled. i'll fix it.

justinmk committed:
snap: more yak-shaving

- generate decryption step
- skip job for PRs
- remove transfer.sh support
- deploy every (master) commit

hurricanehrndz commented on ci/snap/deploy.sh:
```diff
@@ -0,0 +1,19 @@
+
+SNAP=$(find ./ -name "*.snap")
+
+if [[ "$SNAP" =~ "dirty" || "$SNAP" =~ "nightly" ]]; then
```
Drop 14, 16,17,18

justinmk commented on .travis.yml:
```diff
@@ -145,6 +158,35 @@ jobs:
+          - name: transfer
+          - name: lxd
+            channel: stable
+      before_install: /bin/true
```
@hurricanehrndz can this `before_install` be removed?

justinmk commented:
Disabled for now 52981690687429b7c1cc8ac76fbf92560a6420e2 , the deploy step hangs on travis:

https://travis-ci.org/neovim/neovim/builds/618434882

justinmk commented:
Per the [build log](https://travis-ci.org/neovim/neovim/jobs/618447865?utm_medium=notification&utm_source=github_status) we have 2 options:

> If you need to disable confinement, please consider using devmode, but note that devmode revision will only be allowed to be released in edge and beta channels.
Please check the errors and some hints below:
>
>   (NEEDS REVIEW) confinement 'classic' not allowed. If your snap needs classic confinement to function, please make a request for this snap to use classic by creating a new topic in the forum using the 'store-requests' category and detail the technical reasons why classic is required.

Where do we set "devmode"? We could do that for the "edge" channel always, then later we can consider requesting classic mode for "stable" channel.

hurricanehrndz commented:
I have asked for classic confinement, I don't hear a reply within the next
two hours I will make the pull request.

On Fri, Nov 29, 2019 at 2:19 AM Justin M. Keyes <notifications@github.com>
wrote:

> Per the build log
> <https://travis-ci.org/neovim/neovim/jobs/618447865?utm_medium=notification&utm_source=github_status>
> we have 2 options:
>
> If you need to disable confinement, please consider using devmode, but
> note that devmode revision will only be allowed to be released in edge and
> beta channels.
> Please check the errors and some hints below:
>
> (NEEDS REVIEW) confinement 'classic' not allowed. If your snap needs
> classic confinement to function, please make a request for this snap to use
> classic by creating a new topic in the forum using the 'store-requests'
> category and detail the technical reasons why classic is required.
>
> Where do we set "devmode"? We could do that for the "edge" channel always,
> then later we can consider requesting classic mode for "stable" channel.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/neovim/neovim/pull/11428?email_source=notifications&email_token=ABMJBTLO3LZGLB6VWZ34PO3QWDNAXA5CNFSM4JPSAQSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFOKG5I#issuecomment-559719285>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMJBTNTCKN7FYK4AEBRBN3QWDNAXANCNFSM4JPSAQSA>
> .
>


-- 
Regards,

Carlos Hernandez

hurricanehrndz commented:
See #11482
